File: src/api/gitlab.rs
use std::borrow::Cow;
use std::collections::HashMap;

use anyhow::{Result, bail};
use async_trait::async_trait;
use gitlab::GitlabBuilder;
use gitlab::api::merge_requests::MergeRequestState;
use gitlab::api::projects::{jobs, merge_requests, pipelines};
use gitlab::api::{self, groups, projects, users};
use gitlab::api::{AsyncQuery, Pagination};
use serde::{Deserialize, Serialize};

use crate::db::remote_repo::RemoteRepository;
use crate::debug;

use super::{
    Action, Job, JobGroup, JobStatus, ListPullRequestsOptions, PullRequest, PullRequestHead,
    RemoteAPI, RemoteInfo,
};

pub struct GitLab {
    host: String,
    client_builder: GitlabBuilder,
    has_token: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct User {
    username: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Project {
    path: String,
    default_branch: String,
    web_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CreateMergeRequestResponse {
    web_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct MergeRequest {
    iid: u64,
    target_branch: String,
    source_branch: String,
    title: String,
    web_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Pipeline {
    id: u64,
    web_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PipelineJob {
    id: u64,
    name: String,
    stage: String,
    status: String,
    web_url: String,
    commit: PipelineJobCommit,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PipelineJobCommit {
    id: String,
    title: String,
    author_name: String,
    author_email: String,
}

impl GitLab {
    const DEFAULT_HOST: &'static str = "https://gitlab.com";
    const DEFAULT_LIMIT: usize = 100;

    pub fn new(host: &Option<String>, token: &str) -> Self {
        debug!("[gitlab] Building GitLab API client, host: {host:?}, token: {token:?}");
        let host = host.clone().unwrap_or(String::from(Self::DEFAULT_HOST));
        let client_builder = GitlabBuilder::new(&host, token);
        let has_token = !token.is_empty();
        Self {
            host,
            client_builder,
            has_token,
        }
    }
}

#[async_trait]
impl RemoteAPI for GitLab {
    async fn info(&self) -> Result<RemoteInfo> {
        debug!("[gitlab] Getting GitLab API info");
        let client = self.client_builder.build_async().await?;

        let mut auth_user = None;

        let ping = if self.has_token {
            let endpoint = users::CurrentUser::builder().build()?;
            let user: User = endpoint.query_async(&client).await?;
            auth_user = Some(user.username);
            true
        } else {
            super::test_connection(&format!("{}:443", self.host)).await
        };

        let info = RemoteInfo {
            name: Cow::Owned(format!("GitLab: {}", self.host)),
            auth_user,
            ping,
            cache: false,
        };

        debug!("[gitlab] Result: {info:?}");
        Ok(info)
    }

    async fn list_repos(&self, _remote: &str, owner: &str) -> Result<Vec<String>> {
        debug!("[gitlab] Listing repos for: {owner}");
        let client = self.client_builder.build_async().await?;

        let endpoint = groups::projects::GroupProjects::builder()
            .group(owner)
            .build()?;
        let projects: Vec<Project> = api::paged(endpoint, Pagination::Limit(Self::DEFAULT_LIMIT))
            .query_async(&client)
            .await?;
        let repos: Vec<_> = projects.into_iter().map(|p| p.path).collect();
        debug!("[gitlab] Results: {repos:?}");
        Ok(repos)
    }

    async fn get_repo(
        &self,
        remote: &str,
        owner: &str,
        name: &str,
    ) -> Result<RemoteRepository<'static>> {
        debug!("[gitlab] Getting repo: {owner}/{name}");
        let id = format!("{owner}/{name}");
        let client = self.client_builder.build_async().await?;

        let endpoint = projects::Project::builder().project(id).build()?;
        let project: Project = endpoint.query_async(&client).await?;

        let remote_repo = RemoteRepository {
            remote: Cow::Owned(remote.to_string()),
            owner: Cow::Owned(owner.to_string()),
            name: Cow::Owned(name.to_string()),
            default_branch: project.default_branch,
            web_url: project.web_url,
            upstream: None, // TODO: Support upstream for gitlab
            expire_at: 0,
        };

        debug!("[gitlab] Result: {remote_repo:?}");
        Ok(remote_repo)
    }

    async fn create_pull_request(
        &self,
        owner: &str,
        name: &str,
        pr: &PullRequest,
    ) -> Result<String> {
        debug!("[gitlab] Creating pull request: {owner}/{name}, pr: {pr:?}");

        let client = self.client_builder.build_async().await?;
        let id = format!("{owner}/{name}");

        let mut builder = merge_requests::CreateMergeRequest::builder();
        builder.project(id).title(&pr.title).target_branch(&pr.base);
        if let Some(ref body) = pr.body {
            builder.description(body);
        }
        match pr.head {
            PullRequestHead::Branch(ref branch) => {
                builder.source_branch(branch);
            }
            PullRequestHead::Repository(_) => {
                bail!("gitlab does not support cross-repo Merge requests yet");
            }
        }

        // TODO: Allow user to set this
        builder.squash(true);
        builder.remove_source_branch(true);

        let endpoint = builder.build()?;
        let resp: CreateMergeRequestResponse = endpoint.query_async(&client).await?;
        debug!("[gitlab] Response: {resp:?}");
        Ok(resp.web_url)
    }

    async fn list_pull_requests(&self, opts: ListPullRequestsOptions) -> Result<Vec<PullRequest>> {
        debug!("[gitlab] Listing pull requests: {opts:?}");

        let client = self.client_builder.build_async().await?;
        let id = format!("{}/{}", opts.owner, opts.name);

        let mut builder = merge_requests::MergeRequests::builder();
        builder.state(MergeRequestState::Opened).project(id);
        if let Some(id) = opts.id {
            builder.iid(id);
        } else {
            if let Some(head) = opts.head {
                match head {
                    PullRequestHead::Branch(branch) => {
                        builder.source_branch(branch);
                    }
                    PullRequestHead::Repository(_) => {
                        bail!("gitlab does not support cross-repo Merge requests yet");
                    }
                }
            }
            if let Some(base) = opts.base {
                builder.target_branch(base);
            }
        }

        let endpoint = builder.build()?;

        let mrs: Vec<MergeRequest> = api::paged(endpoint, Pagination::Limit(Self::DEFAULT_LIMIT))
            .query_async(&client)
            .await?;
        debug!("[gitlab] MRs: {mrs:?}");

        let mut results = Vec::with_capacity(mrs.len());
        for mr in mrs {
            results.push(PullRequest {
                id: mr.iid,
                base: mr.target_branch,
                head: PullRequestHead::Branch(mr.source_branch),
                title: mr.title,
                web_url: mr.web_url,
                body: None,
            });
        }

        debug!("[gitlab] Results: {results:?}");
        Ok(results)
    }

    async fn get_action_optional(
        &self,
        owner: &str,
        name: &str,
        commit: &str,
    ) -> Result<Option<Action>> {
        debug!("[gitlab] Getting action: {owner}/{name}, commit: {commit}");

        let client = self.client_builder.build_async().await?;
        let id = format!("{owner}/{name}");

        let endpoint = pipelines::Pipelines::builder()
            .project(&id)
            .sha(commit)
            .build()?;
        let mut pipeline: Vec<Pipeline> = endpoint.query_async(&client).await?;
        if pipeline.is_empty() {
            return Ok(None);
        }
        let pipeline = pipeline.remove(0);
        debug!("[gitlab] Pipeline: {pipeline:?}");

        let endpoint = pipelines::PipelineJobs::builder()
            .project(id)
            .pipeline(pipeline.id)
            .build()?;

        let mut pipeline_jobs: Vec<PipelineJob> =
            api::paged(endpoint, Pagination::Limit(Self::DEFAULT_LIMIT))
                .query_async(&client)
                .await?;
        pipeline_jobs.reverse();
        debug!("[gitlab] Pipeline jobs: {pipeline_jobs:?}");

        let mut commit = None;
        let mut message = None;
        let mut user = None;
        let mut email = None;
        let mut stage_indexes: HashMap<String, usize> = HashMap::new();
        let mut job_groups: Vec<JobGroup> = Vec::new();
        for pipeline_job in pipeline_jobs {
            if commit.is_none() {
                commit = Some(pipeline_job.commit.id);
                message = Some(pipeline_job.commit.title);
                user = Some(pipeline_job.commit.author_name);
                email = Some(pipeline_job.commit.author_email);
            }

            let status = match pipeline_job.status.as_str() {
                "created" | "pending" | "waiting_for_resource" => JobStatus::Pending,
                "running" => JobStatus::Running,
                "failed" => JobStatus::Failed,
                "success" => JobStatus::Success,
                "canceled" => JobStatus::Canceled,
                "skipped" => JobStatus::Skipped,
                "manual" => JobStatus::Manual,
                _ => JobStatus::Failed,
            };

            let job = Job {
                id: pipeline_job.id,
                name: pipeline_job.name,
                status,
                web_url: pipeline_job.web_url,
            };
            debug!("[gitlab] Converting job: {job:?}");
            match stage_indexes.get(&pipeline_job.stage) {
                Some(idx) => {
                    job_groups.get_mut(*idx).unwrap().jobs.push(job);
                }
                None => {
                    let group = JobGroup {
                        name: pipeline_job.stage.clone(),
                        web_url: pipeline.web_url.clone(),
                        jobs: vec![job],
                    };
                    let idx = job_groups.len();
                    job_groups.push(group);
                    stage_indexes.insert(pipeline_job.stage, idx);
                }
            }
        }
        if commit.is_none() {
            bail!("no commit info found for action");
        }

        let action = Action {
            web_url: pipeline.web_url,
            commit_id: commit.unwrap(),
            commit_message: message.unwrap(),
            user: user.unwrap(),
            email: email.unwrap(),
            job_groups,
        };
        debug!("[gitlab] Result: {action:?}");
        Ok(Some(action))
    }

    async fn get_job_log(&self, owner: &str, name: &str, id: u64) -> Result<String> {
        debug!("[gitlab] Getting job log: {owner}/{name}, job id: {id}");
        let project = format!("{owner}/{name}");
        let client = self.client_builder.build_async().await?;
        let endpoint = jobs::JobTrace::builder().project(project).job(id).build()?;
        let logs: String = endpoint.query_async(&client).await?;
        debug!("[gitlab] Get job logs done, size: {}", logs.len());
        Ok(logs)
    }
}

#[cfg(test)]
mod tests {
    use std::env;

    use super::*;

    struct TestGitLab {
        client: GitLab,

        user: String,

        group: String,
        repo: String,

        default_branch: String,
        web_url: String,
    }

    fn new() -> Option<TestGitLab> {
        let host = env::var("TEST_GITLAB_HOST").ok()?;
        if host.is_empty() {
            return None;
        }

        let token = env::var("TEST_GITLAB_TOKEN").ok()?;
        if token.is_empty() {
            return None;
        }

        let user = env::var("TEST_GITLAB_USER").ok()?;
        if user.is_empty() {
            return None;
        }

        let group = env::var("TEST_GITLAB_GROUP").ok()?;
        if group.is_empty() {
            return None;
        }

        let repo = env::var("TEST_GITLAB_REPO").ok()?;
        if repo.is_empty() {
            return None;
        }

        let default_branch = env::var("TEST_GITLAB_REPO_DEFAULT_BRANCH").ok()?;
        if default_branch.is_empty() {
            return None;
        }

        let web_url = env::var("TEST_GITLAB_REPO_URL").ok()?;
        if web_url.is_empty() {
            return None;
        }

        Some(TestGitLab {
            client: GitLab::new(&Some(host), &token),
            user,
            group,
            repo,
            default_branch,
            web_url,
        })
    }

    #[tokio::test]
    async fn test_info() {
        let Some(gitlab) = new() else {
            return;
        };
        let info = gitlab.client.info().await.unwrap();
        assert_eq!(
            info,
            RemoteInfo {
                name: Cow::Owned(format!("GitLab: {}", gitlab.client.host)),
                auth_user: Some(gitlab.user.clone()),
                ping: true,
                cache: false,
            }
        );
    }

    #[tokio::test]
    async fn test_list_repos() {
        let Some(gitlab) = new() else {
            return;
        };
        let repos = gitlab
            .client
            .list_repos("gitlab", &gitlab.group)
            .await
            .unwrap();
        assert!(repos.contains(&gitlab.repo));
    }

    #[tokio::test]
    async fn test_get_repo() {
        let Some(gitlab) = new() else {
            return;
        };
        let repo = gitlab
            .client
            .get_repo("gitlab", &gitlab.group, &gitlab.repo)
            .await
            .unwrap();
        assert_eq!(
            repo,
            RemoteRepository {
                remote: Cow::Owned("gitlab".to_string()),
                owner: Cow::Owned(gitlab.group.clone()),
                name: Cow::Owned(gitlab.repo.clone()),
                default_branch: gitlab.default_branch.clone(),
                web_url: gitlab.web_url.clone(),
                upstream: None,
                expire_at: 0,
            }
        );
    }

    #[tokio::test]
    async fn test_pull_request() {
        // TODO: How to test PR?
    }

    #[tokio::test]
    async fn test_action() {
        // TODO: How to test action?
    }

    #[tokio::test]
    async fn test_job_log() {
        // TODO: How to test job logs?
    }
}

File: pkg/remoteapi/gitlab.go
package remoteapi

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"slices"
	"time"

	"github.com/fioncat/roxide/pkg/db"
	gitlab "gitlab.com/gitlab-org/api/client-go"
)

const (
	GitLabHost    = "gitlab.com"
	GitLabVersion = "v4"
)

type GitLab struct {
	client *gitlab.Client

	isPrivate bool
	hasToken  bool

	url string

	limit int
}

func NewGitLab(host, apiURL, token string, limit int, timeout time.Duration) (RemoteAPI, error) {
	var opts []gitlab.ClientOptionFunc
	if apiURL != "" {
		opts = append(opts, gitlab.WithBaseURL(apiURL))
	} else if host != "" {
		apiURL = fmt.Sprintf("https://%s/api/%s", host, GitLabVersion)
		opts = append(opts, gitlab.WithBaseURL(apiURL))
	}

	isPrivate := false
	if apiURL != "" {
		parsed, _ := url.Parse(apiURL)
		isPrivate = parsed.Host != GitLabHost
	} else {
		apiURL = fmt.Sprintf("https://%s/api/%s", GitLabHost, GitLabVersion)
	}

	httpClient := http.Client{Timeout: timeout}
	opts = append(opts, gitlab.WithHTTPClient(&httpClient))

	client, err := gitlab.NewClient(token, opts...)
	if err != nil {
		return nil, fmt.Errorf("create gitlab client: %w", err)
	}

	return &GitLab{
		client:    client,
		isPrivate: isPrivate,
		hasToken:  token != "",
		url:       apiURL,
		limit:     limit,
	}, nil
}

func (g *GitLab) Info() (*RemoteInfo, error) {
	var authOk bool
	if g.hasToken {
		_, _, err := g.client.Users.CurrentUser()
		authOk = err == nil
	}

	_, err := http.Get(g.url)
	ping := err != nil

	name := fmt.Sprintf("GitLab API %s", GitLabVersion)
	if g.isPrivate {
		name = fmt.Sprintf("%s (private)", name)
	}

	return &RemoteInfo{
		Name:   name,
		Auth:   g.hasToken,
		AuthOk: authOk,
		Ping:   ping,
	}, nil
}

func (g *GitLab) ListRepos(owner string) ([]string, error) {
	projects, _, err := g.client.Groups.ListGroupProjects(owner, &gitlab.ListGroupProjectsOptions{
		ListOptions: gitlab.ListOptions{
			PerPage: g.limit,
		},
	})
	if err != nil {
		return nil, err
	}

	names := make([]string, 0, len(projects))
	for _, project := range projects {
		names = append(names, project.Path)
	}

	return names, nil
}

func (g *GitLab) GetRepo(owner, name string) (*RemoteRepository, error) {
	id := fmt.Sprintf("%s/%s", owner, name)
	project, _, err := g.client.Projects.GetProject(id, &gitlab.GetProjectOptions{})
	if err != nil {
		return nil, err
	}

	return &RemoteRepository{
		DefaultBranch: project.DefaultBranch,
		Upstream:      nil,
		WebURL:        project.WebURL,
	}, nil
}

func (g *GitLab) SearchRepos(query string) ([]string, error) {
	projects, _, err := g.client.Search.Projects(query, &gitlab.SearchOptions{
		ListOptions: gitlab.ListOptions{
			PerPage: g.limit,
		},
	})
	if err != nil {
		return nil, err
	}

	names := make([]string, 0, len(projects))
	for _, project := range projects {
		names = append(names, project.NameWithNamespace)
	}

	return names, nil
}

func (g *GitLab) GetMergeRequest(req *MergeRequest) (string, error) {
	if req.Upstream != nil {
		return "", errors.New("now we don't support upstream for gitlab api")
	}

	id := fmt.Sprintf("%s/%s", req.Owner, req.Name)
	mrs, _, err := g.client.MergeRequests.ListProjectMergeRequests(id, &gitlab.ListProjectMergeRequestsOptions{
		State:        db.StringPtr("opened"),
		SourceBranch: db.StringPtr(req.Source),
		TargetBranch: db.StringPtr(req.Target),
	})
	if err != nil {
		return "", err
	}

	if len(mrs) == 0 {
		return "", nil
	}

	return mrs[0].WebURL, nil
}

func (g *GitLab) CreateMergeRequest(req *MergeRequest, title, body string) (string, error) {
	if req.Upstream != nil {
		return "", errors.New("now we don't support upstream for gitlab api")
	}

	id := fmt.Sprintf("%s/%s", req.Owner, req.Name)
	mr, _, err := g.client.MergeRequests.CreateMergeRequest(id, &gitlab.CreateMergeRequestOptions{
		SourceBranch: db.StringPtr(req.Source),
		TargetBranch: db.StringPtr(req.Target),
		Title:        db.StringPtr(title),
		Description:  db.StringPtr(body),
	})
	if err != nil {
		return "", err
	}

	return mr.WebURL, nil
}

func (g *GitLab) GetAction(req *ActionRequest) (*Action, error) {
	var sha *string
	if req.Commit != "" {
		sha = &req.Commit
	}
	var ref *string
	if req.Branch != "" {
		ref = &req.Branch
	}

	id := fmt.Sprintf("%s/%s", req.Owner, req.Name)
	pipelines, _, err := g.client.Pipelines.ListProjectPipelines(id, &gitlab.ListProjectPipelinesOptions{
		SHA: sha,
		Ref: ref,
		ListOptions: gitlab.ListOptions{
			PerPage: g.limit,
		},
	})
	if err != nil {
		return nil, err
	}

	if len(pipelines) == 0 {
		return nil, nil
	}
	pipeline := pipelines[0]

	rawJobs, _, err := g.client.Jobs.ListPipelineJobs(id, pipeline.ID, &gitlab.ListJobsOptions{
		ListOptions: gitlab.ListOptions{
			PerPage: g.limit,
		},
	})
	if err != nil {
		return nil, err
	}
	slices.Reverse(rawJobs)

	var commit *ActionCommit
	stagesIndex := make(map[string]int, len(rawJobs))
	runs := make([]ActionRun, 0, len(rawJobs))

	for _, rawJob := range rawJobs {
		if rawJob.Commit == nil {
			continue
		}
		if commit == nil {
			commit = &ActionCommit{
				ID:          rawJob.Commit.ID,
				Message:     rawJob.Commit.Title,
				AuthorName:  rawJob.Commit.AuthorName,
				AuthorEmail: rawJob.Commit.AuthorEmail,
			}
		} else {
			if commit.ID != rawJob.Commit.ID {
				continue
			}
		}

		status := g.convertStatus(rawJob.Status)
		job := ActionJob{
			ID:     int64(rawJob.ID),
			Name:   rawJob.Name,
			Status: status,
			URL:    rawJob.WebURL,
		}

		stageIndex, ok := stagesIndex[rawJob.Stage]
		if !ok {
			run := ActionRun{
				Name: rawJob.Stage,
				URL:  "",
				Jobs: []ActionJob{job},
			}
			idx := len(runs)
			runs = append(runs, run)
			stagesIndex[rawJob.Stage] = idx
		} else {
			runs[stageIndex].Jobs = append(runs[stageIndex].Jobs, job)
		}
	}

	if commit == nil {
		return nil, errors.New("commit info from GitHub workflow runs is empty")
	}

	return &Action{
		URL:    pipeline.WebURL,
		Commit: *commit,
		Runs:   runs,
	}, nil
}

func (g *GitLab) GetJob(owner, name string, id int64) (*ActionJob, error) {
	pid := fmt.Sprintf("%s/%s", owner, name)
	job, _, err := g.client.Jobs.GetJob(pid, int(id))
	if err != nil {
		return nil, err
	}

	return &ActionJob{
		ID:     int64(job.ID),
		Name:   job.Name,
		Status: g.convertStatus(job.Status),
		URL:    job.WebURL,
	}, nil
}

func (g *GitLab) JobLogs(owner, name string, id int64) (string, error) {
	pid := fmt.Sprintf("%s/%s", owner, name)
	reader, _, err := g.client.Jobs.GetTraceFile(pid, int(id))
	if err != nil {
		return "", err
	}

	data, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("read job logs: %w", err)
	}

	return string(data), nil
}

func (g *GitLab) convertStatus(status string) ActionJobStatus {
	switch status {
	case "created", "pending", "waiting_for_resource":
		return ActionJobPending
	case "running":
		return ActionJobRunning
	case "failed":
		return ActionJobFailed
	case "success":
		return ActionJobSuccess
	case "canceled":
		return ActionJobCanceled
	case "skipped":
		return ActionJobSkipped
	case "manual":
		return ActionJobWaitingForConfirm
	default:
		return ActionJobFailed
	}
}
